formats/json/commonMain/src/kotlinx/serialization/json/JsonNamingStrategy.kt

```diff
@@ -0,0 +1,107 @@
+ *
+ * * Naming strategies do not transform serial names of the types used for the polymorphism, as they always should be specified explicitly.
+ * Values from [JsonClassDiscriminator] or global [JsonBuilder.classDiscriminator] also are not altered.
+ *
+ * ### Controversy about using global naming strategies
+ *
+ * Global naming strategies have one key trait that makes them a debatable and controversial topic:
+ * They are very implicit. It means that by looking only at the definition of the class,
+ * it is impossible to say which names it will have in the serialized form.
+ * As a consequence, naming strategies are not friendly to refactorings. Programmer renaming `myId` to `userId` may forget
+ * to rename `my_id`, and vice versa. Generally, any tools one can imagine work poorly with global naming strategies:
+ * Find Usages/Rename in IDE, full-text search by grep, etc. For them, the original name and the transformed are two different things;
+ * changing one without the other may introduce bugs in many unexpected ways.
+ * The lack of a single place of definition, the inability to use automated tools, and more error-prone code lead
+ * to greater maintenance efforts for code with global naming strategies.
+ * However, there are cases where usage of naming strategies is inevitable, such as interop with existing API or migrating a large codebase.
+ * Therefore, one should carefully weigh the pros and cons before considering adding global naming strategies to an application.
+ */
+@ExperimentalSerializationApi
+public fun interface JsonNamingStrategy {
+    /**
+     * Accepts an original [serialName] (defined by property name in the class or [SerialName] annotation) and returns
+     * a transformed serial name which should be used for serialization and deserialization.
+     *
+     * Besides string manipulation operations, it is also possible to implement transformations that depend on the [descriptor]
+     * and its element (defined by [elementIndex]) currently being serialized.
+     * It is guaranteed that `descriptor.getElementName(elementIndex) == serialName`.
+     * For example, one can choose different transformations depending on [SerialInfo]
+     * annotations (see [SerialDescriptor.getElementAnnotations]) or element optionality (see [SerialDescriptor.isElementOptional]).
+     *
+     * Note that invocations of this function are cached for performance reasons.
+     * Caching strategy is an implementation detail and shouldn't be assumed as a part of the public API contract, as it may be changed in future releases.
+     * Therefore, it is essential for this function to be pure: it should not have any side effects, and it should
+     * return the same String for a given [descriptor], [elementIndex], and [serialName], regardless of the number of invocations.
+     */
+    public fun serialNameForJson(descriptor: SerialDescriptor, elementIndex: Int, serialName: String): String
```

Comment:
The serialName parameter could possibly be renamed to `elementSerialName` as there is also a type serial name. (The documentation here, especially the guarantee in code, make the semantics clear though. The parameter still does present a redundant parameter (as it can be retrieved trivially).